probert/lvm.py

```diff
@@ -0,0 +1,241 @@
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import json
+import os
+import pyudev
+import subprocess
+
+from probert.utils import read_sys_block_size
+
+log = logging.getLogger('probert.lvm')
+
+
+def _lvm_report(cmd, report_key):
+    """ [pvs --reportformat=json -o foo,bar] report_key='pv'
+     {
+         "report": [
+             {
+                 "pv": [
+                     {"pv_name":"/dev/md0", "vg_name":"vg0",
+                      "pv_fmt":"lvm2", "pv_attr":"a--",
+                      "pv_size":"<9.99g", "pv_free":"<6.99g"},
+                     {"pv_name":"/dev/md1", "vg_name":"vg0",
+                      "pv_fmt":"lvm2", "pv_attr":"a--",
+                      "pv_size":"<9.99g", "pv_free":"<9.99g"}
+                 ]
+             }
+         ]
+     }
+    """
+    def _flatten_list(data):
+        return [y for x in data for y in x]
+
+    try:
+        result = subprocess.run(cmd, stdout=subprocess.PIPE,
+                                stderr=subprocess.DEVNULL)
+        output = result.stdout.decode('utf-8')
+    except subprocess.CalledProcessError as e:
+        log.error('Failed to probe LVM devices on system:', e)
+        return None
+
+    if not output:
+        return
+
+    reports = {}
+    try:
+        reports = json.loads(output)
+    except json.decoder.JSONDecodeError as e:
+        log.error('Failed to load LVM json report:', e)
+        return None
+
+    return _flatten_list([report.get(report_key)
+                          for report in reports.get('report', [])
+                          if report_key in report])
+
+
+def probe_pvs_report():
+    return _lvm_report(['pvs', '--reportformat=json'], 'pv')
+
+
+def probe_vgs_report():
+    report_cmd = ['vgs', '--reportformat=json', '--units=B',
+                  '-o', 'vg_name,pv_name,pv_uuid,vg_size']
+    return _lvm_report(report_cmd, 'vg')
+
+
+def probe_lvs_report():
+    return _lvm_report('lvs', 'lv')
+
+
+def dmsetup_info(devname):
+    ''' returns dict of info about device mapper dev.
+
+    {'blkdevname': 'dm-0',
+     'blkdevs_used': 'sdbr,sdbq,sdbp',
+     'lv_name': 'lv_srv',
+     'subsystem': 'LVM',
+     'uuid': 'LVM-lyrZxQgOcgVSlj81LvyUnvq4DW3uLLrfJLI5ieYZR9a2fSOGBK03KM78',
+     'vg_name': 'storage_vg_242x'}
+    '''
+    _SEP = '='
+    fields = (
+        'subsystem,vg_name,lv_name,blkdevname,uuid,blkdevs_used'.split(','))
+    try:
+        output = subprocess.check_output(
+            ['sudo', 'dmsetup', 'info', devname, '-C', '-o',
```

Comment:
Let's not invoke sudo here. Maybe we should have some check and warning if you try to probe as non-root? It's sort of a shame that it doesn't work any more but also inevitable I guess.

Reply:
I can drop this, we don't use it now with the vgs/lvs/pvs report output.

Reply:
That said, we will need some root check; some, but not all storage tools require root;  typically if they're reading root-owned files (like /etc/lvm/* or /etc/mdadm/*).